pkg/target/host.go

```diff
@@ -156,29 +145,99 @@ func (h *Host) Flush(d time.Duration) {
 		case <-t.C:
 			h.Wm.Lock()
 			if h.W != nil {
-				h.W.Flush()
+				err := h.W.Flush()
+				if err != nil {
+					h.closeUpdateHostConnection(updateHostHealthStatus)
+				}
 			}
 			h.Wm.Unlock()
 		}
 	}
 }
 
 // Connect connects to target host via TCP. If unsuccessful, sets conn to nil.
-func (h *Host) Connect() {
-	conn, err := net.DialTimeout("tcp", fmt.Sprintf("%s:%d", h.Name, h.Port),
-		time.Duration(h.ConnTimeoutSec)*time.Second)
+func (h *Host) Connect(updateHostHealthStatus chan *HostStatus, attemptCount int) {
+	dialer := net.Dialer{
+		Timeout:   time.Duration(h.ConnTimeoutSec) * time.Second,
+		KeepAlive: time.Duration(h.KeepAliveSec) * time.Second,
+	}
+	conn, err := dialer.Dial("tcp", fmt.Sprintf("%s:%d", h.Name, h.Port))
 	if err != nil {
 		h.Lg.Warn("connection to host failed",
 			zap.String("host", h.Name),
 			zap.Uint16("port", h.Port))
-		h.Conn = nil
+		h.updateHostConnection(nil)
+		if attemptCount == 1 {
+			h.updateHostHealthStatus(updateHostHealthStatus, false)
+		}
 
 		return
 	}
 
-	h.Conn = conn
+	h.updateHostConnection(conn)
+	h.updateHostHealthStatus(updateHostHealthStatus, true)
 
 	h.Wm.Lock()
 	defer h.Wm.Unlock()
 	h.W = bufio.NewWriterSize(conn, h.bufSize)
 }
+
+func (h *Host) updateHostHealthStatus(updateHostHealthStatus chan *HostStatus, status bool) {
+	updateHostHealthStatus <- &HostStatus{Host: h, Status: status, sigCh: make(chan struct{})}
+}
+
+func (h *Host) updateHostConnection(conn net.Conn) {
+	h.Cm.Lock()
+	defer h.Cm.Unlock()
+	h.Conn = conn
+}
+
+func (h *Host) getHostConnection() net.Conn {
+	h.Cm.RLock()
+	defer h.Cm.RUnlock()
+	return h.Conn
+}
+
+func (h *Host) updateHostConnWriteDeadline(updateHostHealthStatus chan *HostStatus) bool {
+	h.Cm.Lock()
+	defer h.Cm.Unlock()
+	if h.Conn == nil {
+		return false
+	}
+	err := h.Conn.SetWriteDeadline(time.Now().Add(
+		time.Duration(h.SendTimeoutSec) * time.Second))
+	if err != nil {
+		h.Lg.Warn("error setting write deadline. Renewing the connection.",
+			zap.Error(err))
+		h.closeUpdateHostConnection(updateHostHealthStatus)
+		return false
+	}
+	return true
+}
+
+func (h *Host) closeConnection() error {
+	h.Cm.Lock()
+	defer h.Cm.Unlock()
+	if h.Conn != nil {
+		return h.Conn.Close()
+	}
+	return nil
+}
+
+func (h *Host) maintainHostConnection(updateHostHealthStatus chan *HostStatus) {
+	const maxReconnectPeriodMs = 5000
+	const reconnectDeltaMs = 10
+	for {
+		for reconnectWait, attemptCount := 0, 1; h.getHostConnection() == nil; {
+			time.Sleep(time.Duration(reconnectWait) * time.Millisecond)
+			if reconnectWait < maxReconnectPeriodMs {
+				reconnectWait = reconnectWait*2 + reconnectDeltaMs
```

Comment:
Not related to this MR, we should probably randomise reconnectWait

Reply:
Probably, but not sure.

Reply:
Can you give some specific scenarios?

Reply:
Let's have this discussion in different thread.